src/mlpack/base.hpp

```diff
@@ -0,0 +1,124 @@
+
+// First, check if Armadillo was included before, warning if so.
+#ifdef ARMA_INCLUDES
+#pragma message "Armadillo was included before mlpack; this can sometimes cause\
+ problems.  It should only be necessary to include <mlpack/core.hpp> and not \
+<armadillo>."
+#endif
+
+// Defining _USE_MATH_DEFINES should set M_PI.
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+// Next, standard includes.
+#include <cctype>
+#include <cfloat>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+#include <numeric>
+#include <vector>
+#include <queue>
+
+// But if it's not defined, we'll do it.
+#ifndef M_PI
+  #define M_PI 3.141592653589793238462643383279
+#endif
+
+// MLPACK_COUT_STREAM is used to change the default stream for printing
+// purpose.
+#if !defined(MLPACK_COUT_STREAM)
+ #define MLPACK_COUT_STREAM std::cout
+#endif
+
+// MLPACK_CERR_STREAM is used to change the stream for printing warnings
+// and errors.
+#if !defined(MLPACK_CERR_STREAM)
+ #define MLPACK_CERR_STREAM std::cerr
+#endif
+
+// Give ourselves a nice way to force functions to be inline if we need.
+#define force_inline
+#if defined(__GNUG__) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __attribute__((always_inline))
+#elif defined(_MSC_VER) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __forceinline
+#endif
+
+// Backport this functionality from C++14, if it doesn't exist.
+#if __cplusplus <= 201103L
+#if !defined(_MSC_VER) || _MSC_VER <= 1800
+namespace std {
+
+template<bool B, class T = void>
+using enable_if_t = typename enable_if<B, T>::type;
+
+}
+#endif
+#endif
+
+// Backport std::any from C+17 to C++11 to replace boost::any.
+// Use mnmlstc backport implementation only if compiler does not
+// support C++17.
+#if __cplusplus < 201703L
+  #include <mlpack/core/std_backport/any.hpp>
+  #include <mlpack/core/std_backport/string_view.hpp>
+  #define ANY core::v2::any
+  #define ANY_CAST core::v2::any_cast
+  #define STRING_VIEW core::v2::string_view
+#else
+  #include <any>
+  #include <string_view>
+  #define ANY std::any
+  #define ANY_CAST std::any_cast
+  #define STRING_VIEW std::string_view
+#endif 
+
+// Now include Armadillo through the special mlpack extensions.
+#include <mlpack/core/arma_extend/arma_extend.hpp>
+#include <mlpack/core/util/arma_traits.hpp>
+
+// On Visual Studio, disable C4519 (default arguments for function templates)
+// since it's by default an error, which doesn't even make any sense because
+// it's part of the C++11 standard.
+#ifdef _MSC_VER
+  #pragma warning(disable : 4519)
+  #define ARMA_USE_CXX11
+#endif
+
+// Ensure that the user isn't doing something stupid with their Armadillo
+// defines.
+#include <mlpack/core/util/arma_config_check.hpp>
+
+// This can be removed with Visual Studio supports an OpenMP version with
+// unsigned loop variables.
+#ifdef _WIN32
+  #define omp_size_t intmax_t
```

Comment:
This workaround just adds to the maintenance burden, and risks introduction of subtle bugs.  It's simpler and much less hassle to simply disable use of OpenMP if the compiler doesn't support at least OpenMP 3.1 (released in [2011](https://www.openmp.org/specifications/), ie. more than 10 years ago).

I don't think we need to bend over backwards to provide workarounds for the crap that MS turns out.  If they can't provide proper support for at least semi-modern OpenMP, it's really their problem.  People that insist on using windoze can always install a proper compiler via the [mingw64 project](https://www.mingw-w64.org/), or use [WSL2](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) already provided by MS.

Suggested simple check for a minimum supported version of OpenMP:

    #if (defined(_OPENMP) && (_OPENMP >= 201107))
      #undef  MLPACK_USE_OPENMP
      #define MLPACK_USE_OPENMP
    #endif

Then the check elsewhere in mlpack code for OpenMP would be for `MLPACK_USE_OPENMP` instead of `_OPENMP`

Reply:
Agreed.  Long ago I had a lot of idealism about making everything work everywhere, but either I have become less idealistic or I simply don't have enough time for my idealism anymore.  I don't know which is the case here, but either way I agree with your suggestion now, so I applied it in 092c9c91cd.  I think my changes broke something in the build, but I have to run now; I'll fix it later on.